README
Fraucheky - the GPL container for USB devices

							    Version 0.5
							     2017-10-11
						           Niibe Yutaka

What's this?
============

This is a set of routines to deliver GPL with USB device implementation.
It is licensed under GPLv3 (or later).

Please note that I don't know if later version(s) of GPL in future can
be installed on the device (because of its size).

Please see NeuG for a concrete example.


ChangeLog
2017-10-11  NIIBE Yutaka  <gniibe@fsij.org>

	* VERSION: Version 0.5.

	* fraucheky.c (fraucheky_setup_endpoints_for_interface): For
	STM32F103, we don't use DEV.
	* usb-msc.c [GNU_LINUX_EMULATION] (usb_buf_size): Only for
	GNU_LINUX_EMULATION.

2017-10-10  NIIBE Yutaka  <gniibe@fsij.org>

	* build.mk (ARG_BFDNAME): New.
	(BFDNAME_OBJ): Remove.

2017-07-07  NIIBE Yutaka  <gniibe@fsij.org>

	* usb-msc.c (msc_handle_command): Fix command error handling.

2017-07-06  NIIBE Yutaka  <gniibe@fsij.org>

	* build.mk (distclean): Add COPYING.

	* usb-msc.c (msc_handle_command): Handle ATA_16 command
	just like MODE_SENSE6.	Error handling fix.

2017-07-05  NIIBE Yutaka  <gniibe@fsij.org>

	* usb-msc.c [GNU_LINUX_EMULATION] (usb_start_receive): Use
	usb_lld_rx_enable_buf.
	[GNU_LINUX_EMULATION] (EP6_OUT_Callback): No need for rxcpy.
	Use usb_lld_rx_enable_buf.

	* fraucheky.c (fraucheky_setup_endpoints_for_interface): Change
	API.  Support GNU_LINUX_EMULATION.

	* disk-on-rom.c [GNU_LINUX_EMULATION] (msc_scsi_write): Don't
	touch flash.

	* configure (dongdong): Don't output '\'.
	(cluster_map): Output '\' line instead of newline only.

	* build.mk (BFDNAME_OBJ): New.
	(ARG_BFDARCH): New.

2016-06-14  Niibe Yutaka  <gniibe@fsij.org>

	* VERSION: Version 0.4.

2016-06-13  Niibe Yutaka  <gniibe@fsij.org>

	* fraucheky.c, usb-msc.c: Update with new USB API.

2016-06-01  NIIBE Yutaka  <gniibe@fsij.org>

	* fraucheky.c (fraucheky_setup_endpoints_for_interface): Call
	fraucheky_reset to let msc_handle_command return error and
	restart receiving command.

	* usb-msc.c (fraucheky_main_active): New.
	(fraucheky_main_done): Remove.
	* disk-on-rom.c (msc_scsi_stop): Use fraucheky_main_active.

2016-05-30  Niibe Yutaka  <gniibe@fsij.org>

	* disk-on-rom.c: Follow the change of Chopstx SYS API.

	* fraucheky.c: Follow the change of Chopstx USB	API.
	* usb-msc.c (EP6_IN_Callback): Likewise.

2015-07-31  Niibe Yutaka  <gniibe@fsij.org>

	* VERSION: Version 0.3.

2015-07-28  Niibe Yutaka  <gniibe@fsij.org>

	* usb-msc.c (fraucheky_main): Clear fraucheky_main_done at end.
	* disk-on-rom.c (msc_scsi_write): Check fraucheky_enabled.

	* fraucheky.c (fraucheky_setup, fraucheky_get_descriptor): Follow
	the API change of USB driver.

2015-07-27  Niibe Yutaka  <gniibe@fsij.org>

	* configure (check_build_operating_system)
	(get_size_and_timestamp): New.  Support FreeBSD for build system.
	Reported by Kenji Rikitake.

2015-07-26  Kenji Rikitake  <kenji@k2r.org>

	* usb-msc.c (scsi_inquiry_data_83): New.
	(scsi_inquiry_data): Version = 0 (was: 5:SPC-3).
	(msc_handle_command): Better interoperability for SCSI_INQUIRY.

2015-07-25  Niibe Yutaka  <gniibe@fsij.org>

	* usb-msc.c (msc_handle_command): LUN=0 should not be in the LUN
	LIST.  Return empty list (LUN LIST LENGTH = 0).

2015-07-23  Niibe Yutaka  <gniibe@fsij.org>

	* usb-msc.c (msc_handle_command): Handle SCSI_REPORT_LUN command.
	* msc.h (SCSI_REPORT_LUN): New.

2015-03-17  Niibe Yutaka  <gniibe@fsij.org>

	* VERSION: Version 0.2.

	* configure: Generate INDEX substituting @REVISION@.

2015-03-17  Niibe Yutaka  <gniibe@fsij.org>

	* VERSION: Version 0.1a.

2015-01-21  Niibe Yutaka  <gniibe@fsij.org>

	* README: Version 0.1.

	* build.mk (distclean): Add disk-on-rom.h.

	* configure (file_info, cluster_map): New functions.

	* disk-on-rom.c: Include disk-on-rom.h, which is generated.
	(d0_fat0_sector): Use CLUSTER_MAP macro.
	(d0_rootdir_sector): Use macros of COPYING_ATTRIBUTES,
	README_ATTRIBUTES, INDEX_ATTRIBUTES.
	Use macros of COPYING__FILE_SIZE,
	README__FILE_SIZE, INDEX__FILE_SIZE.

2013-11-15  Niibe Yutaka  <gniibe@fsij.org>

	* fraucheky.c (fraucheky_get_descriptor): Handle string descriptor
	request correctly.

2013-11-06  Niibe Yutaka  <gniibe@fsij.org>

	* Version 0.0.



msc.h
#define MSC_CBW_SIGNATURE 0x43425355
#define MSC_CSW_SIGNATURE 0x53425355

#define MSC_CSW_STATUS_PASSED 0
#define MSC_CSW_STATUS_FAILED 1

#define SCSI_INQUIRY                0x12
#define SCSI_MODE_SENSE6            0x1A
#define SCSI_ALLOW_MEDIUM_REMOVAL   0x1E
#define SCSI_READ10                 0x28
#define SCSI_READ_CAPACITY10        0x25
#define SCSI_REQUEST_SENSE          0x03
#define SCSI_START_STOP_UNIT        0x1B
#define SCSI_TEST_UNIT_READY        0x00
#define SCSI_WRITE10                0x2A
#define SCSI_VERIFY10               0x2F
#define SCSI_READ_FORMAT_CAPACITIES 0x23
#define SCSI_SYNCHRONIZE_CACHE      0x35
#define SCSI_REPORT_LUN             0xA0

#define MSC_IDLE        0
#define MSC_DATA_OUT    1
#define MSC_DATA_IN     2
#define MSC_SENDING_CSW 3
#define MSC_ERROR       4

struct CBW {
  uint32_t dCBWSignature;
  uint32_t dCBWTag;
  uint32_t dCBWDataTransferLength;
  uint8_t bmCBWFlags;
  uint8_t bCBWLUN;
  uint8_t bCBWCBLength;
  uint8_t CBWCB[16];
} __attribute__((packed));

struct CSW {
  uint32_t dCSWSignature;
  uint32_t dCSWTag;
  uint32_t dCSWDataResidue;
  uint8_t bCSWStatus;
} __attribute__((packed));

#define SCSI_ERROR_NOT_READY 2
#define SCSI_ERROR_ILLEAGAL_REQUEST 5
#define SCSI_ERROR_UNIT_ATTENTION 6
#define SCSI_ERROR_DATA_PROTECT 7


fraucheky.c
/*
 * flaucheky.c -- the USB Mass Storage Class device
 *
 * Copyright (C) 2013, 2015, 2016, 2017
 *               Free Software Initiative of Japan
 * Author: NIIBE Yutaka <gniibe@fsij.org>
 *
 * This file is a part of Fraucheky, making sure to have GNU GPL on a
 * USB thumb drive
 *
 * Fraucheky is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * Fraucheky is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdint.h>
#include <stdlib.h>
#include "usb_lld.h"
#include "config.h"

/* MSC BULK_IN, BULK_OUT */
/* EP6: 64-byte, 64-byte */
#define ENDP6_TXADDR        (0x180)
#define ENDP6_RXADDR        (0x1c0)

#define USB_INITIAL_FEATURE 0x80   /* bmAttributes: bus powered */

#define MSC_GET_MAX_LUN_COMMAND        0xFE
#define MSC_MASS_STORAGE_RESET_COMMAND 0xFF

/* USB Standard Device Descriptor */
static const uint8_t device_desc[] = {
  18,   /* bLength */
  DEVICE_DESCRIPTOR,     /* bDescriptorType */
  0x10, 0x01,   /* bcdUSB = 1.1 */
  0x00,   /* bDeviceClass: 0 means deferred to interface */
  0x00,   /* bDeviceSubClass */
  0x00,   /* bDeviceProtocol */
  0x40,   /* bMaxPacketSize0 */
#include "fraucheky-vid-pid-ver.c.inc"
  1, /* Index of string descriptor describing manufacturer */
  2, /* Index of string descriptor describing product */
  3, /* Index of string descriptor describing the device's serial number */
  1  /* bNumConfigurations */
};

#define MSC_TOTAL_LENGTH (9+7+7)

/* Configuation Descriptor */
static const uint8_t config_desc[] = {
  9,			         /* bLength: Configuation Descriptor size */
  CONFIG_DESCRIPTOR,             /* bDescriptorType: Configuration */
  (9+9+7+7), 0x00,               /* wTotalLength:no of returned bytes */
  1,				 /* bNumInterfaces: */
  0x01,                          /* bConfigurationValue: Configuration value */
  0x00,				 /* iConfiguration.  */
  USB_INITIAL_FEATURE,		 /* bmAttributes*/
  50,				 /* MaxPower 100 mA */

  /* Interface Descriptor.*/
  9,			         /* bLength: Interface Descriptor size */
  INTERFACE_DESCRIPTOR,          /* bDescriptorType: Interface         */
  0,				 /* bInterfaceNumber.                  */
  0x00,				 /* bAlternateSetting.                 */
  0x02,				 /* bNumEndpoints.                     */
  0x08,				 /* bInterfaceClass (Mass Stprage).    */
  0x06,				 /* bInterfaceSubClass (SCSI
				    transparent command set, MSCO
				    chapter 2).                        */
  0x50,				 /* bInterfaceProtocol (Bulk-Only
				    Mass Storage, MSCO chapter 3).     */
  0x00,				 /* iInterface.                        */
  /* Endpoint Descriptor.*/
  7,			         /* bLength: Endpoint Descriptor size  */
  ENDPOINT_DESCRIPTOR,   	 /* bDescriptorType: Endpoint          */
  0x86,				 /* bEndpointAddress: (IN6)            */
  0x02,				 /* bmAttributes (Bulk).               */
  0x40, 0x00,			 /* wMaxPacketSize.                    */
  0x00,				 /* bInterval (ignored for bulk).      */
  /* Endpoint Descriptor.*/
  7,			         /* bLength: Endpoint Descriptor size  */
  ENDPOINT_DESCRIPTOR,   	 /* bDescriptorType: Endpoint          */
  0x06,				 /* bEndpointAddress: (OUT6)           */
  0x02,				 /* bmAttributes (Bulk).               */
  0x40, 0x00,			 /* wMaxPacketSize.                    */
  0x00,				 /* bInterval (ignored for bulk).      */
};


/* USB String Descriptors */
static const uint8_t string_lang_id[] = {
  4,				/* bLength */
  STRING_DESCRIPTOR,
  0x09, 0x04			/* LangID = 0x0409: US-English */
};

#include "fraucheky-usb-strings.c.inc"

static const uint8_t string_serial[] = {
  8*2+2,			/* bLength */
  STRING_DESCRIPTOR,    	/* bDescriptorType */
  /* Serial number: "FSIJ-0.0" */
  'F', 0, 'S', 0, 'I', 0, 'J', 0, '-', 0, '0', 0, '.', 0, '0', 0,
};


struct desc { const uint8_t *desc; uint16_t size; };

static const struct desc string_descriptors[] = {
  {string_lang_id, sizeof (string_lang_id)},
  {string_vendor, sizeof (string_vendor)},
  {string_product, sizeof (string_product)},
  {string_serial, sizeof (string_serial)},
};

void
fraucheky_setup_endpoints_for_interface (struct usb_dev *dev, int stop)
{
  extern void fraucheky_reset (void);

  if (!stop)
    {
#ifdef GNU_LINUX_EMULATION
      usb_lld_setup_endp (dev, ENDP6, 1, 1);
#else
      (void)dev;
      usb_lld_setup_endpoint (ENDP6, EP_BULK, 0, ENDP6_RXADDR, ENDP6_TXADDR, 64);
#endif
      fraucheky_reset ();
    }
  else
    {
      usb_lld_stall_tx (ENDP6);
      usb_lld_stall_rx (ENDP6);
    }
}

int
fraucheky_setup (struct usb_dev *dev)
{
  struct device_req *arg = &dev->dev_req;

  static const uint8_t lun_table[] = { 0, 0, 0, 0, };

  if (USB_SETUP_GET (arg->type))
    {
      if (arg->request == MSC_GET_MAX_LUN_COMMAND)
	return usb_lld_ctrl_send (dev, lun_table, sizeof (lun_table));
    }
  else /* SETUP_SET */
    if (arg->request == MSC_MASS_STORAGE_RESET_COMMAND)
      return usb_lld_ctrl_ack (dev);

  return -1;
}

int
fraucheky_get_descriptor (struct usb_dev *dev)
{
  struct device_req *arg = &dev->dev_req;
  uint8_t rcp = arg->type & RECIPIENT;
  uint8_t desc_type = (arg->value >> 8);
  uint8_t desc_index = (arg->value & 0xff);

  if (rcp == DEVICE_RECIPIENT)
    {
      if (desc_type == DEVICE_DESCRIPTOR && arg->index == 0)
	return usb_lld_ctrl_send (dev, device_desc, sizeof (device_desc));
      else if (desc_type == CONFIG_DESCRIPTOR && arg->index == 0)
	return usb_lld_ctrl_send (dev, config_desc, sizeof (config_desc));
      else if (desc_type == STRING_DESCRIPTOR)
	{
	  if (desc_index >= sizeof (string_descriptors) / sizeof (struct desc)
	      || !((arg->index == 0 && desc_index == 0)
		   || arg->index == 0x0409))
	    /* We only provide string in English.  */
	    return -1;

	  return usb_lld_ctrl_send (dev,
				    string_descriptors[desc_index].desc,
				    string_descriptors[desc_index].size);
	}
    }

  return -1;
}


disk-on-rom.c
/*
 * disk-on-rom.c -- FAT storage (GPL, README, and INDEX.HTM) on ROM
 *
 * Copyright (C) 2013, 2015, 2016, 2017
 *               Free Software Initiative of Japan
 * Author: NIIBE Yutaka <gniibe@fsij.org>
 *
 * This file is a part of Fraucheky, GNU GPL in a USB thumb drive
 *
 * Fraucheky is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * Fraucheky is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdint.h>
#include <string.h>
#include <chopstx.h>

#include "disk-on-rom.h"
#include "msc.h"
#include "sys.h"

extern int fraucheky_main_active;
extern int fraucheky_enabled (void);

extern uint8_t _binary_COPYING_start;
extern uint8_t _binary_COPYING_end;
extern uint8_t _binary_README_start;
extern uint8_t _binary_README_end;
extern uint8_t _binary_INDEX_start;
extern uint8_t _binary_INDEX_end;

int (*p_msc_scsi_write) (uint32_t lba, const uint8_t *buf, size_t size);
int (*p_msc_scsi_read) (uint32_t lba, const uint8_t **sector_p);
void (*p_msc_scsi_stop) (uint8_t code);

#define TOTAL_SECTORS 128	/* Maximum 64KB.  */
#define SECTOR_SIZE 512

/*
 * blk=0: master boot record sector
 * blk=1: fat0
 * blk=2: fat1
 * blk=3: root directory
 * blk=4: fat cluster #2
 * ...
 * blk=4+123: fat cluster #2+123
 */

static const uint8_t d0_0_sector[] = {
  0xeb, 0x3c,             /* Jump instruction */
  0x90,                   /* NOP instruction */
  0x6d, 0x6b, 0x64, 0x6f, 0x73, 0x66, 0x73, 0x00, /* "mkdosfs" */
  0x00, 0x02,             /* Bytes per sector: 512 */
  0x01,                   /* sectors per cluster: 1 */
  0x01, 0x00,             /* reserved sector count: 1 */
  0x02,                   /* Number of FATs: 2 */
  0x10, 0x00,             /* Max. root directory entries: 16 (1 sector) */
  TOTAL_SECTORS, 0x00,    /* total sectors: 128 */
  0xf8,                   /* media descriptor: fixed disk */
  0x01, 0x00,             /* sectors per FAT: 1 */
  0x20, 0x00,             /* sectors per track: 32 */
  0x40, 0x00,             /* number of heads: 64 */
  0x00, 0x00, 0x00, 0x00, /* hidden sectors: 0 */
  0x00, 0x00, 0x00, 0x00, /* total sectors (long) */
  0x00,                   /* drive number */
  0x00,                   /* reserved */
  0x29,                   /* extended boot signature */
  0x9e, 0x5a, 0x2b, 0x68, /* Volume ID (serial number) (Little endian) */
  /* Volume label: Fraucheky */
  'F', 'r', 'a', 'u', 'c', 'h', 'e', 'k', 'y', ' ', ' ',
  0x46, 0x41, 0x54, 0x31, 0x32, 0x20, 0x20, 0x20, /* FAT12 */
  0x0e,                   /*    push cs */
  0x1f,                   /*    pop ds */
  0xbe, 0x5b, 0x7c,       /*    mov si, offset message_txt */
  0xac,                   /* 1: lodsb */
  0x22, 0xc0,             /*    and al, al */
  0x74, 0x0b,             /*    jz 2f */
  0x56,                   /*    push si */
  0xb4, 0x0e,             /*    mov ah, 0eh */
  0xbb, 0x07, 0x00,       /*    mov bx, 0007h */
  0xcd, 0x10,             /*    int 10h ; output char color=white */
  0x5e,                   /*    pop si */
  0xeb, 0xf0,             /*    jmp 1b */
  0x32, 0xe4,             /* 2: xor ah, ah */
  0xcd, 0x16,             /*    int 16h; key input */
  0xcd, 0x19,             /*    int 19h; load OS */
  0xeb, 0xfe,             /* 3: jmp 3b */
  /* "This is not a bootable disk... \r\n" */
  0x54, 0x68, 0x69, 0x73, 0x20,
  0x69, 0x73, 0x20, 0x6e, 0x6f, 0x74, 0x20, 0x61,
  0x20, 0x62, 0x6f, 0x6f, 0x74, 0x61, 0x62, 0x6c,
  0x65, 0x20, 0x64, 0x69, 0x73, 0x6b, 0x2e, 0x20,
  0x20, 0x50, 0x6c, 0x65, 0x61, 0x73, 0x65, 0x20,
  0x69, 0x6e, 0x73, 0x65, 0x72, 0x74, 0x20, 0x61,
  0x20, 0x62, 0x6f, 0x6f, 0x74, 0x61, 0x62, 0x6c,
  0x65, 0x20, 0x66, 0x6c, 0x6f, 0x70, 0x70, 0x79,
  0x20, 0x61, 0x6e, 0x64, 0x0d, 0x0a, 0x70, 0x72,
  0x65, 0x73, 0x73, 0x20, 0x61, 0x6e, 0x79, 0x20,
  0x6b, 0x65, 0x79, 0x20, 0x74, 0x6f, 0x20, 0x74,
  0x72, 0x79, 0x20, 0x61, 0x67, 0x61, 0x69, 0x6e,
  0x20, 0x2e, 0x2e, 0x2e, 0x20, 0x0d, 0x0a, 0x00,
};


/* Sector no for Root:       3 */
#define DROPHERE_SECTOR      4
#define COPYING_SECTOR_START 5
#define COPYING_SECTOR_END   (COPYING_SECTOR_START+COPYING_BLOCKS-1)
#define README_SECTOR_START  (COPYING_SECTOR_END+1)
#define README_SECTOR_END    (README_SECTOR_START+README_BLOCKS-1)
#define INDEX_SECTOR_START   (README_SECTOR_END+1)
#define INDEX_SECTOR_END     (INDEX_SECTOR_START+INDEX_BLOCKS-1)

#define CLSTR_NO(sec_no) (sec_no-2)

static const uint8_t d0_fat0_sector[] = {
  0xf8, 0xff, 0xff,  /* Media descriptor: fixed disk *//* EOC */
  CLUSTER_MAP
};

static const uint8_t d0_rootdir_sector[] = {
  'F', 'r', 'a',  'u',  'c',  'h',  'e',  'k',  'y',  ' ',  ' ', 
  /* "Fraucheky  " */
  0x08, /* Volume label */
  0x00,
  0x00, 0xe7, 0x63, 0x65, 0x43, /* Create Time */
  0x65, 0x43,  /* last access */
  0x00, 0x00,
  0xe7, 0x63, 0x65, 0x43,  /* last modified */
  0x00, 0x00, /* cluster # */
  0x00, 0x00, 0x00, 0x00, /* file size */

  'C', 'O', 'P',  'Y',  'I',  'N',  'G',  ' ',  ' ',  ' ',  ' ', 
  /* "COPYING     " */
  COPYING_ATTRIBUTES,
  CLSTR_NO (COPYING_SECTOR_START), 0x00,  /* cluster # */
  COPYING_FILE_SIZE,

  'R', 'E', 'A',  'D',  'M',  'E',  ' ',  ' ',  ' ',  ' ',  ' ', 
  /* "README      " */
  README_ATTRIBUTES,
  CLSTR_NO (README_SECTOR_START), 0x00,  /* cluster # */
  README_FILE_SIZE,

  'I', 'N', 'D', 'E', 'X', ' ', ' ', ' ', 'H', 'T', 'M', /* INDEX.HTM */
  INDEX_ATTRIBUTES,
  CLSTR_NO (INDEX_SECTOR_START), 0x00, /* cluster # */
  INDEX_FILE_SIZE,

  'D', 'R', 'O', 'P', 'H', 'E', 'R', 'E', ' ', ' ', ' ', /* DROPHERE */
  0x10, /* Sub directory */
  0x00,
  0x64, 0x10, 0x64, 0x65, 0x43,  /* Create Time */
  0x63, 0x43, /* last access */
  0x00, 0x00,
  0xe4, 0x74, 0x16, 0x43, /* last modified */
  CLSTR_NO (DROPHERE_SECTOR), 0x00,  /* cluster # */
  0x00, 0x00, 0x00, 0x00  /* file size */
};

static const uint8_t d0_drophere_sector[] = {
  '.', ' ', ' ',  ' ',  ' ',  ' ',  ' ',  ' ',  ' ',  ' ',  ' ', 
  /* ".          " */
  0x10, /* directory */
  0x00,
  0x00, 0xe7, 0x63, 0x65, 0x43, /* Create Time */
  0x65, 0x43,  /* last access */
  0x00, 0x00,
  0xe7, 0x63, 0x65, 0x43,  /* last modified */
  CLSTR_NO (DROPHERE_SECTOR), 0x00, /* cluster # */
  0x00, 0x00, 0x00, 0x00, /* file size */

  '.', '.', ' ',  ' ',  ' ',  ' ',  ' ',  ' ',  ' ',  ' ',  ' ', 
  /* "..         " */
  0x10, /* directory */
  0x00,
  0x00, 0xe7, 0x63, 0x65, 0x43, /* Create Time */
  0x65, 0x43,  /* last access */
  0x00, 0x00,
  0xe7, 0x63, 0x65, 0x43,  /* last modified */
  0x00, 0x00, /* cluster # */
  0x00, 0x00, 0x00, 0x00, /* file size */
};

static uint8_t the_sector[512];

const uint16_t rom_var = { 0xffff };

int
msc_scsi_write (uint32_t lba, const uint8_t *buf, size_t size)
{
  if (p_msc_scsi_write)
    return (*p_msc_scsi_write) (lba, buf, size);

#if !defined(GNU_LINUX_EMULATION)
  if (fraucheky_enabled () && lba == DROPHERE_SECTOR)
    {
      flash_unlock ();
      flash_program_halfword ((uintptr_t)&rom_var, 0);
    }
#endif

  return 0;
}

int
msc_scsi_read (uint32_t lba, const uint8_t **sector_p)
{
  uint32_t offset;
  uint32_t size;

  if (p_msc_scsi_read)
    return (*p_msc_scsi_read) (lba, sector_p);

  if (lba >= TOTAL_SECTORS)
    return SCSI_ERROR_ILLEAGAL_REQUEST;

  *sector_p = the_sector;

  switch (lba)
    {
    case 0:			/* MBR */
      memcpy (the_sector, d0_0_sector, sizeof d0_0_sector);
      memset (the_sector + sizeof d0_0_sector, 0, 512 - sizeof d0_0_sector);
      the_sector[510] = 0x55;
      the_sector[511] = 0xaa;
      return 0;

    case 1:
    case 2:			/* FAT */
      memcpy (the_sector, d0_fat0_sector, sizeof d0_fat0_sector);
      memset (the_sector + sizeof d0_fat0_sector, 0,
	      512 - sizeof d0_fat0_sector);
      return 0;

    case 3:			/* Root directory.  */
      memcpy (the_sector, d0_rootdir_sector, sizeof d0_rootdir_sector);
      memset (the_sector + sizeof d0_rootdir_sector, 0,
	      512 - sizeof d0_rootdir_sector);
      return 0;

    case 4:			/* DROPHERE directory.  */
      memcpy (the_sector, d0_drophere_sector, sizeof d0_drophere_sector);
      memset (the_sector + sizeof d0_drophere_sector, 0,
	      512 - sizeof d0_drophere_sector);
      return 0;

    default:
      if (lba >= COPYING_SECTOR_START && lba <= COPYING_SECTOR_END)
	{
	  offset = (lba - COPYING_SECTOR_START) * SECTOR_SIZE;
	  size = SECTOR_SIZE;

	  if (lba == COPYING_SECTOR_END)
	    {
	      size = &_binary_COPYING_end - &_binary_COPYING_start
		- (COPYING_BLOCKS - 1) * SECTOR_SIZE;
	      if (size != SECTOR_SIZE)
		memset (the_sector + size, 0, SECTOR_SIZE - size);
	    }

	  memcpy (the_sector, &_binary_COPYING_start + offset, size);
	}
      else if (lba >= README_SECTOR_START && lba <= README_SECTOR_END)
	{
	  offset = (lba - README_SECTOR_START) * SECTOR_SIZE;
	  size = SECTOR_SIZE;

	  if (lba == README_SECTOR_END)
	    {
	      size = &_binary_README_end - &_binary_README_start
		- (README_BLOCKS - 1) * SECTOR_SIZE;
	      if (size != SECTOR_SIZE)
		memset (the_sector + size, 0, SECTOR_SIZE - size);
	    }

	  memcpy (the_sector, &_binary_README_start + offset, size);
	}
      else if (lba >= INDEX_SECTOR_START && lba <= INDEX_SECTOR_END)
	{
	  offset = (lba - INDEX_SECTOR_START) * SECTOR_SIZE;
	  size = SECTOR_SIZE;

	  if (lba == INDEX_SECTOR_END)
	    {
	      size = &_binary_INDEX_end - &_binary_INDEX_start
		- (INDEX_BLOCKS - 1) * SECTOR_SIZE;
	      if (size != SECTOR_SIZE)
		memset (the_sector + size, 0, SECTOR_SIZE - size);
	    }

	  memcpy (the_sector, &_binary_INDEX_start + offset, size);
	}
      else
	memset (the_sector, 0, 512);
      return 0;
    }
}

void
msc_scsi_stop (uint8_t code)
{
  if (p_msc_scsi_stop)
    (*p_msc_scsi_stop) (code);
  else
    fraucheky_main_active = 0;
}
